### PR TITLE
Refactor Status tag styles into UI component #8656 

### PR DESCRIPTION
**Pull request by salty-ivy**
-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [x] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]

addresses issue #8656 
-added a new status_tag.html as UI component 
-using inclusion tag created a new tag called status_tag , args( page, classname )
-created a status_tag.stories.tsx for ```LIVE```  and ```DRAFT``` variations

### DIFFERENCE ### 
as in  case if page is not live it still renders a span in an anchor tag like the live condition 
but with disabled css class
### OUTPUT ### 

**New test **
image

**Styleguide**

image

**Story**

ezgif-5-67376ee7a6

### CODE REVIEW

**Comment by lb-** [comment]
Thanks @salty-ivy - have tagged this for review.

File: wagtail/admin/templates/wagtailadmin/shared/status_tag.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/status_tag.html:5]: Let's pull out this shared bit into a `fragment`, here's an example of this in use

https://github.com/wagtail/wagtail/blob/main/wagtail/admin/templates/wagtailadmin/panels/inline_panel_child.html#L4

**Review by lb- — CHANGES_REQUESTED** [review]
@salty-ivy  great stuff here, however, I think we need to rethink this a bit.

You have pulled out part of the page_status_tag shared include into a new `shared/status_tag` with a template tag of `status_tag`.

However, this generic `status_tag` now is very much tied to the requirement of a page and it's status.

Instead, I think the original goal of #8656 was to solve the generic use case of just status tags - not just page status'. I know that this can be a bit confusing, but I think let's try to solve the simplest scenarios first and get to 80% of the status tags using this new template tag.

From that PR, once merged, we can revisit the page status tag afresh and see if there is a way to make that also use the new generic status tag.

Also, by doing this, we kind of have to make some assumptions about the `a` & `span` tag variants, which would be good to avoid until later.

I recommend we put this PR to a draft and if you are up for it, give this another go with just tackling the generic case first.

See https://github.com/wagtail/wagtail/issues/8656#issuecomment-1345658164 for more details.

**Comment by salty-ivy** [comment]
> @salty-ivy  great stuff here, however, I think we need to rethink this a bit.
> 
> 
> 
> You have pulled out part of the page_status_tag shared include into a new `shared/status_tag` with a template tag of `status_tag`.
> 
> 
> 
> However, this generic `status_tag` now is very much tied to the requirement of a page and it's status.
> 
> 
> 
> Instead, I think the original goal of #8656 was to solve the generic use case of just status tags - not just page status'. I know that this can be a bit confusing, but I think let's try to solve the simplest scenarios first and get to 80% of the status tags using this new template tag.
> 
> 
> 
> From that PR, once merged, we can revisit the page status tag afresh and see if there is a way to make that also use the new generic status tag.
> 
> 
> 
> Also, by doing this, we kind of have to make some assumptions about the `a` & `span` tag variants, which would be good to avoid until later.
> 
> 
> 
> I recommend we put this PR to a draft and if you are up for it, give this another go with just tackling the generic case first.
> 
> 
> 
> See https://github.com/wagtail/wagtail/issues/8656#issuecomment-1345658164 for more details.
> 
> 
> 
> 

Ahh, I see it now pardon me for the trouble 😅 I got really confused with we were actually trying to do and completely focused on logical extraction of page_status_tag now I see how it is , it was little confusing at first at page_status_tag seemed pretty much refactored to me 
Sorry for the trouble I would like to keep working on this one

**Comment by salty-ivy** [comment]
@lb-  how about 
1. using a kwarg in our tag to tell whether to use link or span depending on the condition something like
```{% status  use="a"%}``` or  ```{% status use="span" %}``` to specify the what to render 
### alternatively ### 
putting  ```<a></a>```  and  ```<span></span>```  into fragments and and taking up kwarg to specify which fragment to use?

File: wagtail/admin/templates/wagtailadmin/shared/status_tag.html
Comment (RealOrangeOne) [inline wagtail/admin/templates/wagtailadmin/shared/status_tag.html:4]: Issue: This should definitely use the `{% pageurl %}` tag for improved performance.

File: wagtail/admin/templates/wagtailadmin/shared/status_tag.stories.tsx
Comment (RealOrangeOne) [inline wagtail/admin/templates/wagtailadmin/shared/status_tag.stories.tsx:13]:
```suggestion
      options: [null, 'https://wagtail.org'],
```

**Comment by lb-** [comment]
@salty-ivy  I really think it may not be worth solving for page_status_tag yet - it will introduce a potentially complex convention of outer tags if we use `use='a'` and it's probably not something we need right now.

Solving for the simple status tag will cover 80% of the usage of `status-tag` class reuse - giving us a nice way to build on this in the future if we need to refactor classes or refine the way these are used.

The page status being a span or a link depending on context is probably not a pattern we would support long term, I think in the future it would be more likely we would go with a different approach here.

**Comment by salty-ivy** [comment]
Got it

File: wagtail/admin/templates/wagtailadmin/shared/status_tag.html
Comment (salty-ivy) [reply wagtail/admin/templates/wagtailadmin/shared/status_tag.html]: alright, i will putting it in use

File: wagtail/admin/templates/wagtailadmin/shared/status_tag.html
Comment (salty-ivy) [reply wagtail/admin/templates/wagtailadmin/shared/status_tag.html]: i think lets not put it in the partial at all ,  putting only ```span``` part of status that has been used through out 
putting it here without any further plan of action might create confusion like ```page_status_tag``` and ```page_status_tag_new```

File: wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html
Comment (zerolab) [inline wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html:4]: The comment about using `pageurl` instead of `page.url` is defeated by this if

How about something like:

```jinja2
{% pageurl page as the_page_url %}
{% if the_page_url is not None %}
   <a href="{{ the_page_url }} ...
{% else %}
 ...
{% endif %}
```

? (note: did not test locally, but should work)

Comment (salty-ivy) [reply wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html:4]: Ahh yes I didn't realize I used already used, still working on it would get everything in prefer

**Comment by salty-ivy** [comment]
@lb-  i have made the changes added all stories  tests updated the templates for the tag, and did some error handling as well
however backend circleCI is still failing because of some other issues, which i believe has not originated from the changes i made.

rest everything is ready to be reviewed again thanks you

File: wagtail/admin/templates/wagtailadmin/pages/workflow_history/detail.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/pages/workflow_history/detail.html:29]: I'm not sure this will work as is. Did you happen to validate this manually?

File: wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html:16]: Might be good to keep the existing nesting here - will make the diff cleaner

File: wagtail/snippets/tests/test_snippets.py
Comment (lb-) [inline wagtail/snippets/tests/test_snippets.py:4054]: Best to avoid changing whitespace in other files unless really needed.

**Review by lb- — COMMENTED** [review]
A few quick comments - I have not done a full review though. 

Mostly looks good (thank you!) - the workflow one seems broken though.

File: wagtail/admin/templates/wagtailadmin/shared/status_tag.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/status_tag.html:1]: Just check your editor settings - looks like your files are being saved without a line break  at the end.

Comment (salty-ivy) [reply wagtail/admin/templates/wagtailadmin/shared/status_tag.html:1]: no
actually that caused an error in tests

Comment (salty-ivy) [reply wagtail/admin/templates/wagtailadmin/shared/status_tag.html:1]: image
 snippet's test was looking for exact one line html while status tag was injecting the one with line breaks

Comment (salty-ivy) [reply wagtail/admin/templates/wagtailadmin/shared/status_tag.html:1]: so i had to put the status tag partial's html in one to pass the test

File: wagtail/admin/templates/wagtailadmin/pages/workflow_history/detail.html
Comment (salty-ivy) [reply wagtail/admin/templates/wagtailadmin/pages/workflow_history/detail.html:29]: well the error was quite simple that  blocktrans doesn't allow any other block tag causing and error, this approached didn't seems to cause any error in the tests
image
in wagtail.admin.tests.pages.test_workflow_history.TestWorkflowHistoryDetail

File: wagtail/admin/templates/wagtailadmin/shared/status_tag.html
Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/shared/status_tag.html:1]: Ok - you can use spaceless to solve most of this problem. 

https://docs.djangoproject.com/en/4.1/ref/templates/builtins/#spaceless

We will also want to add some docs to the template file so maybe look at the icon template html is done as a reference (it also accepts classname).

File: wagtail/admin/templates/wagtailadmin/pages/workflow_history/detail.html
Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/pages/workflow_history/detail.html:29]: Ok. 

But it looks like the variable `workflow_state.get_status_display` is not used correctly

Should it be `{{ workflow_state.get_status_display }}` instead?

Comment (salty-ivy) [reply wagtail/admin/templates/wagtailadmin/pages/workflow_history/detail.html]: yes !

File: wagtail/admin/templates/wagtailadmin/shared/status_tag.html
Comment (salty-ivy) [reply wagtail/admin/templates/wagtailadmin/shared/status_tag.html]: i just checked through the snippet test its still passing even with the line break, but there's been a multiple cases where status tag has been used in single line between other tags , i am little confused about it should we still use ``` spaceless```?

Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/shared/status_tag.html]: Avoid spaceless unless you need it. And if it's only one or two tests that fail - sometimes it's best to refine the test assertion to not be so strict.

Comment (salty-ivy) [reply wagtail/admin/templates/wagtailadmin/shared/status_tag.html]: alright then its good to go there's no test failing on that scope

**Comment by salty-ivy** [comment]
ready to be reviewed .

**Comment by gasman** [comment]
Sorry if I'm jumping in on a detail that's already been decided, but are we definitely sure that an opening/closing tag pair is the right pattern for this, rather than a single tag like `{% status "Disabled" classname="primary" %}`? I know it's replacing an opening/closing `<span>` tag, but it's not really being used as a container for content in the same way that, say, a header element is. I'm concerned that we're just replicating the underlying HTML structure here, rather than serving as a proper abstraction layer for it.

**Comment by salty-ivy** [comment]
@gasman you have a point , i am not sure but this was originally decided considering its future scope.

this block based approach has  potential  to eventually take up other ```status-tag``` css for example ``` <span class="visuallyhidden"></span>``` is spread across the admin just like this one, also there are some tags with ```<a></a>```
eg. ```page_status_tag``` and ```page_status_tag_new``` which can't be covered in a single tag approach, these potentially can also be included in this one

**Comment by gasman** [comment]
I don't think the presence of `<a>` tags would prevent the single tag approach - the template tag implementation would just become something like:

    {% if url %}
        <a href="{{ url }}" class="status-tag {{ classname }}">{{ label }}</a>
    {% else %}
        <span class="status-tag {{ classname }}">{{ label }}</span>
    {% endif %}

and that way, the template tag is serving an additional useful purpose, by ensuring that the logic for "if a URL is passed, use an `<a>` tag, otherwise use a `<span>`" only needs to exist in one place.

**Comment by salty-ivy** [comment]
I am not really sure @lb- would have a better idea on this one 
As far as I understood it was more about abstracting the css rather than the html.

**Comment by salty-ivy** [comment]
> @salty-ivy  I really think it may not be worth solving for page_status_tag yet - it will introduce a potentially complex convention of outer tags if we use `use='a'` and it's probably not something we need right now.
> 
> 
> 
> Solving for the simple status tag will cover 80% of the usage of `status-tag` class reuse - giving us a nice way to build on this in the future if we need to refactor classes or refine the way these are used.
> 
> 
> 
> The page status being a span or a link depending on context is probably not a pattern we would support long term, I think in the future it would be more likely we would go with a different approach here.

.

**Comment by salty-ivy** [comment]
@lb-  this PR is ready, although we might need some discussion on @gasman's comment let me know if there's  anything that needs to be done

**Comment by lb-** [comment]
@salty-ivy - Matt's idea above is a great one, it gives us the benefit of keeping the status template tag generic but supports the `a` tag without getting too complicated.

We will need to do a few things though

1. if provided with a `url` render the `a` variant, but also in that case handle an optional `title` variable
2. Hard-code the target blank & rel noreferrer though for now

**Comment by salty-ivy** [comment]
Alright, I would make the changes soon

**Comment by salty-ivy** [comment]
in this case i think inclusion_tag must be used to register the tag along with its arguments
in cases where there's another span or may be trans blocks inside of the ```status``` tag i think we need to use the ``` fragment ``` tag to extract the html and then pass it as a variable in the ```status``` tag.
eg:-
```
<a href="{{page_url}}" target="_blank" rel="noreferrer" class="status-tag primary" title="{% trans 'Visit the live page' %}">
    <span class="visuallyhidden">{% trans "Current page status:" %}</span> {{ page.status_string }}
</a>
```
would become 

```
{% fragment as trans_title %}{% trans 'Visit the live page' %}{% endfragment %}
        {% fragment as inner_span %}<span class="visuallyhidden">{% trans "Current page status:" %}</span> {{ page.status_string }}{% endfragment %}
        {% status inner_span url=page_url title=trans_title classname="primary" %}
```

is there any other compact alternative ? or we can go with it

**Comment by gasman** [comment]
Translated strings should be set up with `{% trans ... as foo %}`, not a fragment.

IMO, any time that we find ourselves needing to insert new elements like `<span class="visuallyhidden">` inside the tag body, the `{% status %}` tag implementation should be extended to support that as a built-in feature:

    {% trans 'Visit the live page' as trans_title %}
    {% trans 'Current page status' as trans_label %}
    {% status url=page_url title=trans_title hidden_label=trans_label classname="primary" %}

(as an aside... if we find ourselves having to use `trans` on tag attributes all the time, maybe we could make a template filter for that?

    {% status url=page_url title="Visit the live page"|trans hidden_label="Current page status"|trans classname="primary" %}

Surely that must exist already...)

Otherwise, we're never really getting away from having to think "what does the internal HTML structure of this element look like?" when invoking the `{% status %}` tag, and it's not really serving its purpose of abstracting away the HTML, only adding another layer of complexity.

**Comment by salty-ivy** [comment]
@gasman i see it now , getting the correct conventions thanks for the clearance

**Comment by lb-** [comment]
@salty-ivy let me know if you need another review on this. Thanks for working through the other feedback.

**Comment by salty-ivy** [comment]
> @salty-ivy let me know if you need another review on this. Thanks for working through the other feedback. 

I would make few remaining changes and request for the review soon

**Comment by salty-ivy** [comment]
> @salty-ivy let me know if you need another review on this. Thanks for working through the other feedback.

@lb-  itr ready for the review

**Comment by salty-ivy** [comment]
@lb-  now its ready for the review

**Comment by lb-** [comment]
Thanks @salty-ivy

**Comment by salty-ivy** [comment]
converting to draft since it got conflicts, would resolve those soon

File: wagtail/admin/templates/wagtailadmin/pages/history.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/pages/history.html:41]: `status_label` might be a better variable name here and in other places.

File: wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html:13]: Similar here - maybe `status_hidden_label`

File: wagtail/admin/templates/wagtailadmin/shared/status_tag.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/status_tag.html:2]:
```suggestion
    <a href="{{url}}" target="_blank" rel="noreferrer" class='{% classnames "status-tag" classname %}' {% if title %}title="{{title}}"{% endif %}>
```

Using the template util is much easier `classnames`

File: wagtail/admin/templates/wagtailadmin/workflows/task_index.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/workflows/task_index.html:40]: We should use translate values for this and the above 'disabled' labels

File: wagtail/admin/templates/wagtailadmin/shared/status_tag.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/status_tag.html:10]: this line seems overly convoluted, I think we can change it to

```suggestion
    <span class="{% classnames 'status-tag' classname %}">
        {% if hidden_label %}<span class="visuallyhidden">{{hidden_label}}</span>{% endif %}
        {% if label %}{{ label }}{% endif %}
    </span>
```

**Review by lb- — COMMENTED** [review]
@salty-ivy  thanks for this, I should probably help you out with learning how to rebase at some point - DM me in slack.

I have rebased and resolved all the conflicts, applied a few changes as per the comments and will see how the PR goes with the CI now. (note: it will probably fail formatting/testing - but we can clean up easily).

Please review and let me know if there is anything missed, I have a backup of your branch here. https://github.com/lb-/wagtail/tree/temp-pr-9721

I will endeavour to help get this finalised and merged in.

As for Matt's comment above about the translation filter, I think that is a great idea but I think we should do that as a separate issue / PR (there may be a package that exists already also).

Great stuff on this, thanks for approaching this a few different ways, it's pretty close to ready.

**Comment by lb-** [comment]
Will have to solve for URL not being opened in a blank tab - will commit a few extra things as a new commit on top.

**Comment by salty-ivy** [comment]
Great! Thanks a lot for the help will give this a look, once I am done with stimulus issue, let me know if there's anything that I need to do.

**Review by lb- — APPROVED** [review]
looking good, I have had to clean up a few extra things and some missing translation usages but did a smoke test of a whole bunch of tags and working well.

I have pushed up the changes and will check on the CI, if all good we will merge in.

File: wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html:7]: @salty-ivy  I have refactored your original approach a bit, the `url` usage was not always needing to set target/rel so I added the ability to pass through an `attrs`.

It's a tough balance between supporting all usage explicitly (e.g. `url` added does variant A or B) but I opted for this approach as it would have made the template tag more complicated if we needed a `open_in_new_tab` type option.

File: wagtail/admin/templates/wagtailadmin/tables/status_tag_cell.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/tables/status_tag_cell.html:3]: @salty-ivy  there were a few leftover `div` usage items so I have converted them to the status tag also, not sure if these were there already but we can emulate a div by `span` with `w-block` override in the class.

File: wagtail/admin/templatetags/wagtailadmin_tags.py
Comment (lb-) [inline wagtail/admin/templatetags/wagtailadmin_tags.py:789]: @salty-ivy  adding a trailing comma on longer lists (e.g. params) will make the formatter (black) wrap to multiple lines. There is no hard rule but after about 3 or 4 items it reads better if they are all on their own line.

File: wagtail/admin/templatetags/wagtailadmin_tags.py
Comment (salty-ivy) [reply wagtail/admin/templatetags/wagtailadmin_tags.py]: yes, i was originally confused Cause backend was failing something like ..`no attribute called get_host()` but by the looks of it
it was something related some property wasn't passing through the templates (that's what i could assumed) so i tried taking context, @lb-  if you got any idea about that? would like to know

Comment (lb-) [reply wagtail/admin/templatetags/wagtailadmin_tags.py]: ahh yep - I am seeing that now also. Investigation for tomorrow :)

**Comment by lb-** [comment]
ok - still a few more issues - will take another look tomorrow / this week.

**Comment by salty-ivy** [comment]
i couldn't really figure out `client object has no attribute get_host()..`

File: wagtail/admin/templates/wagtailadmin/shared/status_tag.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/status_tag.html:3]:
```suggestion
    <a href="{{ url }}" class="{% classnames 'status-tag' classname %}"{% if title %} title="{{ title }}" {% endif %}{% if attrs %} {{ attrs }}{% endif %}>
```

File: wagtail/admin/templates/wagtailadmin/tables/status_tag_cell.html
Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/tables/status_tag_cell.html:3]: Note. I think I'm wrong here. The inline-block on the base styles would be all we need.

File: wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html:3]: @salty-ivy  it was this line that was causing the error - note that in the original code the `page_url` was declared and wrapped in `with`.

I think this probably also caused a bit of a problem with other changes in this file when merging `main` in earlier.

I see the goal here of using the `pageurl`  template tag but will lean towards leaving the previous `with` statement in place.

I will push up the changes to this file and see how we go. Will also do another round of testing.

**Comment by lb-** [comment]
@gasman it possible can you do another quick read through, as I have changed a few items and rebased since your last comments.

Thanks for the epic stuff here @salty-ivy - hope to get this merged in soon.

File: wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html
Comment (salty-ivy) [reply wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html:3]: ahhhh i seee, @lb-  thank you so much, for finding the root cause, i wasn't able to find out what's actually causing it, should have looked harder !!!sorry for the trouble.

Comment (zerolab) [reply wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html:3]: note: please use the `pageurl` tag (`{% pageurl page as page_url %}` as it make use of the request and benefit from cached site parts and more

Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html:3]: If it's ok @zerolab I'd like to get that in as a seperate PR - unless I've missed something that's not being used now in this file. 

When it was being used it caused some other tests to fail so it probably is worth its own PR.

I'm happy to ensure that gets done once this PR is merged.

Comment (zerolab) [reply wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html:3]: Works for me! Will look at submitting a separate PR that does it across the admin, not just the one template

Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/shared/page_status_tag.html:3]: Issue raised https://github.com/wagtail/wagtail/issues/10125

**Comment by lb-** [comment]
https://github.com/wagtail/wagtail/issues/10126 raised from the comment above about the trans filter.

**Comment by lb-** [comment]
I'll do another smoke test and probably merge this in a bit later if all is well

**Comment by lb-** [comment]
Follow on things to do

* Update classes to be `w-status` and use BEM prefixes

Review additional `status-tag` usage and migrate to the template tag if possible
* `wagtail/admin/templates/wagtailadmin/collections/_privacy_switch.html` (may not be able to as it's a button) - maybe this should not be styled as a status tag but instead be styled as a small button - similar to wagtail/admin/templates/wagtailadmin/shared/side_panels/includes/status/privacy.html
* `wagtail/admin/templates/wagtailadmin/reports/workflow.html`
* `wagtail/admin/templates/wagtailadmin/shared/workflow_history/detail.html`
* `wagtail/admin/templates/wagtailadmin/tables/status_flag_cell.html`
* `wagtail/contrib/redirects/templates/wagtailredirects/list.html`
* `wagtail/admin/templates/wagtailadmin/tables/status_flag_cell.html`

**Comment by gasman** [comment]
Sorry I didn't get chance to look at this again before it was merged, but very happy with the updates 👍

**Pull request merged**
